### PR TITLE
fix(bulk-model-sync-gradle): fix MPS Runner configuration for server …

### PR DESCRIPTION
**Pull request by odzhychko**
…sources with base revision

When a base revision was specified, the specified branch was not passed as a system property to the bulk sync.

When a base revision was specified without a source revision, the property value for the source revision was the string `"null"`.

To write test with TestKit we would need further adjustments to the build setup. For this fix, I considered them out of scope.

I tried to test execution with Gradles TestKit first. See https://docs.gradle.org/current/userguide/test_kit.html But the Plugin tries to download the package "org.modelix:bulk-model-sync-mps" as `mpsDependencies`. The names of the downloaded files are then used as `classPathElements` in the MPS Runner configuration. Therefore, the plugin cannot be tested with TestKit without publishing "org.modelix:bulk-model-sync-mps" and its dependents locally or including the "modelix.core" project from within a test execution.

## To be verified by reviewers

* [ ] Relevant public API members have been documented
* [ ] Documentation related to this PR is complete
  * [ ] Boundary conditions are documented
  * [ ] Exceptions are documented
  * [ ] Nullability is documented if used
* [ ] Touched existing code has been extended with documentation if missing
* [ ] Code is readable
* [ ] New features and fixed bugs are covered by tests

### CODE REVIEW

File: bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        val expectedJvmArgs = listOf(
+            "-Dmodelix.mps.model.sync.bulk.input.path=/jsonDir",
+            "-Dmodelix.mps.model.sync.bulk.input.modules=",
+            "-Dmodelix.mps.model.sync.bulk.input.modules.prefixes=",
+            "-Dmodelix.mps.model.sync.bulk.repo.path=/repositoryDir",
+            "-Dmodelix.mps.model.sync.bulk.input.continueOnError=false",
+            "-Xmx2g",
+        )
+
+        assertEquals(expectedJvmArgs, jvmArgs)
```
Comment (languitar) [inline bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt:33]: Maybe use set comparison because the argument order probably doesn't matter?

Comment (odzhychko) [reply bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt]: Yeah, that would be cleaner.
I changed it to Kotest to use the appropriate assertions.

**Review by languitar — COMMENTED** [review]
I can't judge the other changes as I haven't seen much about how MPS is launched so far.

File: bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        jvmArgs shouldContainExactlyInAnyOrder expectedJvmArgs
+    }
+
+    @Test
+    fun `build configuration for local target with a server source with a base revision and branch name`() {
+        val serverSource = ServerSource(
+            url = "aUrl",
+            repositoryId = "aRepositoryId",
+            branchName = "aBranchName",
+            baseRevision = "aBaseRevision",
```
Comment (mhuster23) [inline bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt:42]: We also need similar test cases for `baseRevision=null`, as it is also valid to provide a revision or pair of repoId and branchName without a baseRevision.

These cases are broken right now with this PRs changes.
The fact that this isn't covered by existing tests right now emphasizes our need to be able to test multiple gradle configurations.

**Review by mhuster23 — CHANGES_REQUESTED** [review]
Small side note: From a reviewer's perspective it would have been nice to separate the actual fix and the refactoring into separate commits. It took me a bit to spot the fix.

File: bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt (hunk 2 above)
Comment (odzhychko) [reply bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt:42]: > as it is also valid to provide a revision or pair of repoId and branchName without a baseRevision.

If a `baseRevision` is not provided, the `repositoryId` and `branchName` are not passed to the `MPSBulkSynchronizer`. The data is downloaded in a separate MPS task.

So that case should not be broken.

> The fact that this isn't covered by existing tests right now emphasizes our need to be able to test multiple gradle configurations.

Definitely. But for now, this is out of scope for me.

**Comment by odzhychko** [comment]
> Small side note: From a reviewer's perspective it would have been nice to separate the actual fix and the refactoring into separate commits. It took me a bit to spot the fix.

I will do that.

File: bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt (hunk 2 above)
Comment (mhuster23) [reply bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt:42]: > If a baseRevision is not provided, the repositoryId and branchName are not passed to the MPSBulkSynchronizer.

Ah yes, you are right it is only relevant for the `ExportFromModelServerTask`.

> But for now, this is out of scope for me.

Agreed.

**Comment by odzhychko** [comment]
> Small side note: From a reviewer's perspective it would have been nice to separate the actual fix and the refactoring into separate commits. It took me a bit to spot the fix.

Done

File: bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt (hunk 2 above)
Comment (odzhychko) [reply bulk-model-sync-gradle/src/test/kotlin/org/modelix/model/sync/bulk/gradle/MpsRunnerConfigurationTest.kt:46]: > If a baseRevision is not provided, the repositoryId and branchName are not passed to the MPSBulkSynchronizer.

I update the test case `build configuration for local target with a server source without a base revision` to test this, just to be sure.

**Review by mhuster23 — DISMISSED** [review]

**Review by mhuster23 — APPROVED** [review]
